source/extensions/transport_sockets/tls/ocsp/ocsp.h

```diff
@@ -0,0 +1,324 @@
+ */
+enum class CertStatus {
+  // The certificate is known to be valid
+  GOOD,
+  // The certificate has been revoked
+  REVOKED,
+  // The responder has no record of the certificate and cannot confirm its validity
+  UNKNOWN
+};
+
+/**
+ * Reflection of the ASN.1 CertId structure.
+ * Contains the information to uniquely identify an SSL Certificate.
+ * Serial numbers are guaranteed to be
+ * unique per issuer but not necessarily universally.
+ */
+struct CertId {
+  CertId(std::string serial_number, std::string alg_oid, std::string issuer_name_hash,
+         std::string issuer_public_key_hash);
+
+  std::string serial_number_;
+  std::string alg_oid_;
+  std::string issuer_name_hash_;
+  std::string issuer_public_key_hash_;
+};
+
+/**
+ * Reflection of the ASN.1 SingleResponse structure.
+ * Contains information about the OCSP status of a single certificate.
+ * An OCSP request may request the status of multiple certificates and
+ * therefore responses may contain multiple SingleResponses.
+ *
+ * this_update_ and next_update_ reflect the validity period for this response.
+ * If next_update_ is not present, the OCSP responder always has new information
+ * available. In this case the response would be considered immediately expired
+ * and invalid for stapling.
+ */
+struct SingleResponse {
+  SingleResponse(CertId cert_id, CertStatus status, Envoy::SystemTime this_update,
+                 absl::optional<Envoy::SystemTime> next_update);
+
+  const CertId cert_id_;
+  const CertStatus status_;
+  const Envoy::SystemTime this_update_;
+  const absl::optional<Envoy::SystemTime> next_update_;
+};
+
+/**
+ * Reflection of the ASN.1 ResponseData structure.
+ * Contains an OCSP response for each certificate in a given request
+ * as well as the time at which the response was produced.
+ */
+struct ResponseData {
+  ResponseData(Envoy::SystemTime produced_at, std::vector<SingleResponse> single_responses);
+
+  const Envoy::SystemTime produced_at_;
+  const std::vector<SingleResponse> single_responses_;
+};
+
+/**
+ * An abstract type for OCSP response formats. Which variant of |Response| is
+ * used in an |OcspResponse| is indicated by the structure's OID.
+ *
+ * We currently enforce that OCSP responses must be for a single certificate
+ * only. The methods on this class extract the relevant information for the
+ * single certificate contained in the response.
+ */
+class Response {
+public:
+  virtual ~Response() = default;
+
+  /**
+   * @return The number of certs reported on by this response.
+   */
+  virtual size_t getNumCerts() PURE;
+
+  /**
+   * @return The revocation status of the certificate.
+   */
+  virtual CertStatus getCertRevocationStatus() PURE;
+
+  /**
+   * @return The serial number of the certificate.
+   */
+  virtual const std::string& getCertSerialNumber() PURE;
+
+  /**
+   * @return The beginning of the validity window for this response.
+   */
+  virtual const Envoy::SystemTime& getThisUpdate() PURE;
```

Comment:
Does this name come from an RFC? If not, can you rename to `beginningValidityWindow()` and the next one to `endingValidityWindow()`?

Reply:
these do come directly from the RFC

Reply:
Ok, probably best to leave them as-is then.